README.md
# ng5-slider
[![npm version](https://badge.fury.io/js/ng5-slider.svg)](https://badge.fury.io/js/ng5-slider)
[![Travis CI Build](https://travis-ci.org/angular-slider/ng5-slider.svg?branch=master)](https://travis-ci.org/angular-slider/ng5-slider)

Website: https://angular-slider.github.io/ng5-slider/

Self-contained, mobile friendly slider component for Angular 5+ based on [angularjs-slider](https://github.com/angular-slider/angularjs-slider).

## Demos

 * Single slider - [StackBlitz](https://stackblitz.com/edit/ng5-slider-simple-slider-example?file=src%2Fapp%2Fapp.component.ts)

   ![simple slider image](https://raw.githubusercontent.com/angular-slider/ng5-slider/master/assets/simple-slider.png)

 * Range slider - [StackBlitz](https://stackblitz.com/edit/ng5-slider-range-slider-example?file=src%2Fapp%2Fapp.component.ts)

   ![range slider image](https://raw.githubusercontent.com/angular-slider/ng5-slider/master/assets/range-slider.png)

 * Slider with ticks - [StackBlitz](https://stackblitz.com/edit/ng5-slider-ticks-example?file=src%2Fapp%2Fapp.component.ts)

   ![ticks slider image](https://raw.githubusercontent.com/angular-slider/ng5-slider/master/assets/ticks-slider.png)

 * Customised slider - [StackBlitz](https://stackblitz.com/edit/ng5-slider-customised-range-slider-example?file=src%2Fapp%2Fapp.component.ts)

   ![customised slider image](https://raw.githubusercontent.com/angular-slider/ng5-slider/master/assets/customised-slider.png)

 * Slider with custom style - [StackBlitz](https://stackblitz.com/edit/ng5-slider-styled-slider-example?file=src%2Fapp%2Fapp.component.ts)

   ![styled slider image](https://raw.githubusercontent.com/angular-slider/ng5-slider/master/assets/styled-slider.png)

 * Vertical slider - [StackBlitz](https://stackblitz.com/edit/ng5-slider-vertical-slider-example?file=src%2Fapp%2Fapp.component.ts)

   ![vertical slider image](https://raw.githubusercontent.com/angular-slider/ng5-slider/master/assets/vertical-slider.png)

 * [More examples on official website](https://angular-slider.github.io/ng5-slider/demos)

## Dependencies

 * Angular 5+
 * rxjs 5+

## Installation

To add the slider to your Angular project:
```
npm install --save ng5-slider
```

Once installed, add the slider to your `app.module.ts`:
```typescript
import { Ng5SliderModule } from 'ng5-slider';

...

@NgModule({
   ...
   imports: [
     ...
     Ng5SliderModule,
    ...
   ],
   ...
})
export class AppModule {}
```

## Sample usage

Now you can use the slider component in your app components, for example in `app.component.ts`:
```typescript
import { Options } from 'ng5-slider';
...

@Component({...})
export class AppComponent {
  value: number = 100;
  options: Options = {
    floor: 0,
    ceil: 200
  };
}
```

And in template file `app.component.html`:
```html
<ng5-slider [(value)]="value" [options]="options"></ng5-slider>
```

## Documentation
Full API documentation is available on [official website](https://angular-slider.github.io/ng5-slider/docs).

## Styling

An overview of how to apply your own style to the slider is described in [STYLING.md](STYLING.md).

## Animations

As of v1.2.0 the slider features CSS animations of slider movement. If you prefer the previous behaviour, without animations, you can set the flag `animate: false` in your slider options.

## Keyboard Shortcuts

In addition to mouse/touch events, the slider can also be controlled through keyboard. The available shortcuts are:
 - right/up arrow - increase by single step,
 - left/down arrow - decrease by single step,
 - page up - increase by 10% of slider range,
 - page down - decrease by 10% of slider range,
 - home - move to minimum value,
 - end - move to maximum value.

## Tooltips

The slider allows for customising how to implement tooltips. See [TOOLTIPS.md](TOOLTIPS.md) for more information.

## Known Issues

Before reporting a new bug, please look at [KNOWN_ISSUES.md](KNOWN_ISSUES.md) for a list of known problems and their workarounds. New bugs reports for these problems will not be accepted.

## Bugs

You can report any bugs as [Github issues](https://github.com/angular-slider/ng5-slider/issues).

Please describe the issue in detail pasting any relevant code, or preferrably a StackBlitz with reproduction of the problem by [forking and editing this sample StackBlitz](https://stackblitz.com/edit/ng5-slider-simple-slider-example?file=src/app/app.component.ts). Please also provide the version of NPM package you are using.

## Changelog

For list of changes and bugfixes, see [CHANGELOG.md](CHANGELOG.md).

## Developer information

If you would like to contribute to the project, see [DEVELOPERS.md](DEVELOPERS.md).

## License

The project is licensed under the MIT license.


CHANGELOG.md
# 1.2.6
 * re-publish with correct README.md, LICENSE.md and scss files
 * fix build script to avoid publishing mistakes in the future
 * document available keyboard shortcuts in README.md (issue #218)

# 1.2.5
 * fix rxjs imports to work correctly with Angular 9+ (issues #181 and #197)
 * fix ngOnDestroy throwing an exception (issue #215)
 * fix "Attempt to use a destroyed view" error on Angular 8+ (issue #140)
 * allow slider to be used with OnPush change detection strategy (issue #158)
 * remove unnecessary packages from package.json (issue #113)
 * update some outdated dependencies as reported by dependabot

# 1.2.4
 * fix bug with setting vertical state on slider initialisation

# 1.2.3
 * fix visibility of high slider pointer in case of inherited CSS style

# 1.2.2
 * fix detecting nearest slider handle when clicking on slider bar (issue #110)
 * finish refactoring legacy slider code to modern Angular 2+ style

# 1.2.1
 * fix erratic slider behaviour when using simple slider with ticksArray
 * fix tick parameters not getting updated correctly (issue #105 and #89)
 * fix not getting highValue in userChange events (issue #108)
 * fix slider not getting refreshed after visibility change - see notes in KNOWN_ISSUES.md (issue #106)

# 1.2.0
 * add CSS animations ported from angularjs-slider (issue #72)
 * re-write model update code with rxjs, fixing a number of bugs in this area (issue #69, #86, #91, #93, #95)
 * fix behaviour of draggable range when moving it with mouse (issue #80)
 * fix prepare script on non-Unix platforms (issue #103)

# 1.1.14
 * add ability to set focus programmatically (issue #64)
 * clarify interaction of stepsArray with other options (issue #60)
 * document known issues in README
 * remove obsolete enforceRange option
 * remove some redudant refresh and change callbacks

# 1.1.13
 * fix regression when updating model values through reactive forms (issues #61, #62)

# 1.1.12
 * fix regression caused by applying model value and options changes in the wrong order when both change at the same time (issue #56)
 * allow dynamic HTML in ticks legend (issue #49)
 * improve rendering performance of simple sliders (PR #59)

# 1.1.11
 * don't normalise values when stepsArray option is used (issue #52)

# 1.1.10
 * always convert input values to numbers
 * make sure range invariant is always satisfied (issue #48 follow-up)

# 1.1.9
 * make sure model values are always in allowed range (issue #48)

# 1.1.8
 * fix broken link in README (PR #44)
 * add documentation for changing options object dynamically
 * adjust layout of API documentation page on mobile

# 1.1.7
 * add support for Angular reactive forms (issue #24)
 * fix Angular peer dependency (issue #39)
 * add CSS styling guide
 * embed API documentation in demo app

# 1.1.6
 * fix event handling when clicking on tick value (issue #36)
 * fix warnings about event listener violation (issue #22)
 * new layout of demo app

# 1.1.5
 * fix event handling when touching exactly the tick element (issue #32)
 * update model value if value is not initially in steps array (issue #35)
 * added more e2e tests

# 1.1.4
 * fixed slider jumping to first value if value is not in steps array (issue #29)
 * fixed floating-point rounding in slider value calculations (issue #25)
 * exposed SCSS files in distribution package (PR #28)

# 1.1.3
 * changed initialization code to fix issue with slider shown dynamically through CSS animations (issue #16)

# 1.1.2
 * added manual refresh mechanism to the slider (issue #15)

# 1.1.1
 * fixed AOT compilation bug (issue #13)
 * added support for user-initiated events (issue #9)
 * minor demo app changes

# 1.1.0
 * removed dependency on ng-bootstrap, providing an HTML standard or user-customisable way of rendering tooltips (issue #7)

# 1.0.1
 * moved repository to official angular-slider organisation
 * updated demo app to run under older browsers

# 1.0.0
 * fixed error related to TouchEvent constructor on Firefox and Safari (issue #4)
 * downgraded Angular and ng-bootstrap dependency versions to the minimum required (v5.0.2 and v1.0.0 respectively)
 * fixed errors when using slider component without ng-bootstrap installed
 * added e2e tests for slider component
 * added Travis CI integration

# 0.4.0
 * refactored demo app to show the examples with code snippets
 * reverted view encapsulation of slider component to Angular defaults
 * removed duplication in package.json and README.md between the main project and the NPM library
 * added more API documentation
 * added generation of API documentation using typedoc
 * added scripts to publish demo app and API documentation to Github Pages
 * updated README.md and added DEVELOPERS.md

# 0.3.2
 * add Plunker example to README

# 0.3.1
 * add README.md and LICENSE to NPM package

# 0.3.0
 * re-brand to ng5-slider as the target Angular version is 5+
 * create and release NPM package


# 0.2.0
 * add more examples of usage
 * fix slider behaviour where it was broken in the examples
 * change slider options following refactoring and code design considerations
 * complete first iteration of code refactoring:
   - finish adding TypeScript type signatures everywhere
   - start extracting code to smaller files
   - rename some cryptic variables and functions to more descriptive names
   - start adding Angular bindings for CSS styles currently managed by JqLiteWrapper

# 0.1.0

 * rough rewrite of original code into Angular 2
   - copy-paste of the original JS code
   - add TypeScript type signatures where known
   - adapt from angularjs jqLite element wrappers to a wrapper based on Angular 2's Renderer2


src/app/app.component.ts
import { Component } from '@angular/core';
import { Ng2SliderOptions } from './ng2-slider/ng2-slider.module';
import { LabelType, CustomStepDefinition } from './ng2-slider/options';

interface SimpleSliderModel {
  value: number;
  options: Ng2SliderOptions;
}

interface RangeSliderModel {
  minValue: number;
  maxValue: number;
  options: Ng2SliderOptions;
}

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent {
  simpleSlider: SimpleSliderModel = {
    value: 200,
    options: {
      floor: 0,
      ceil: 500
    }
  };

  rangeSlider: RangeSliderModel = {
    minValue: 10,
    maxValue: 90,
    options: {
      floor: 0,
      ceil: 100,
       step: 10
    }
  };

  styledSlider: RangeSliderModel = {
    minValue: 10,
    maxValue: 90,
    options: {
      floor: 0,
      ceil: 100,
      step: 10,
      showTicks: true
    }
  };

  minMaxLimitSlider: SimpleSliderModel = {
    value: 50,
    options: {
      floor: 0,
      ceil: 100,
      step: 1,
      minLimit: 10,
      maxLimit: 90
    }
  };

  minMaxRangeSlider: RangeSliderModel = {
    minValue: 10,
    maxValue: 90,
    options: {
      floor: 0,
      ceil: 100,
      step: 1,
      minRange: 10,
      maxRange: 50
    }
  };

  noSwitchingRangeSlider: RangeSliderModel = {
    minValue: 10,
    maxValue: 90,
    options: {
      floor: 0,
      ceil: 100,
      step: 1,
      noSwitching: true
    }
  };

  minMaxAndPushRangeSlider: RangeSliderModel = {
    minValue: 60,
    maxValue: 70,
    options: {
      floor: 0,
      ceil: 100,
      step: 1,
      minRange: 10,
      maxRange: 30,
      pushRange: true
    }
  };

  selectionBarSlider: SimpleSliderModel = {
    value: 10,
    options: {
      showSelectionBar: true
    }
  };

  selectionBarAtEndSlider: SimpleSliderModel = {
    value: 5,
    options: {
      floor: 0,
      ceil: 10,
      showSelectionBarEnd: true
    }
  };

  selectionBarFromValueSlider: SimpleSliderModel = {
    value: 5,
    options: {
      floor: -10,
      ceil: 10,
      showSelectionBarFromValue: 0
    }
  };

  selectionBarWithGradientSlider: RangeSliderModel = {
    minValue: 0,
    maxValue: 80,
    options: {
      ceil: 100,
      showSelectionBar: true,
      selectionBarGradient: {
        from: 'white',
        to: '#FC0'
      }
    }
  };

  selectionBarWithDynamicColorSlider: SimpleSliderModel = {
    value: 12,
    options: {
      showSelectionBar: true,
      getSelectionBarColor: (value: number): string => {
        if (value <= 3) {
            return 'red';
        }
        if (value <= 6) {
            return 'orange';
        }
        if (value <= 9) {
            return 'yellow';
        }
        return '#2AE02A';
      }
    }
  };

  dynamicPointerColorSlider: SimpleSliderModel = {
    value: 12,
    options: {
        showSelectionBar: true,
        getPointerColor: (value: number): string => {
            if (value <= 3) {
                return 'red';
            }
            if (value <= 6) {
                return 'orange';
            }
            if (value <= 9) {
                return 'yellow';
            }
            return '#2AE02A';
        }
    }
  };

  floorCeilStepSlider: SimpleSliderModel = {
    value: 12,
    options: {
      floor: 10,
      ceil: 100,
      step: 5
    }
  };

  rtlSlider: SimpleSliderModel = {
    value: 20,
    options: {
      floor: 10,
      ceil: 100,
      step: 5,
      rightToLeft: true
    }
  };

  floatSlider: SimpleSliderModel = {
    value: 0.5,
    options: {
      floor: 0,
      ceil: 2,
      step: 0.1,
      precision: 1
    }
  };

  customDisplayFunctionSlider: RangeSliderModel = {
    minValue: 100,
    maxValue: 400,
    options: {
      floor: 0,
      ceil: 500,
      translate: (value: number): string => {
        return '$' + value;
      }
    }
  };

  customCombineLabelsFunctionSlider: RangeSliderModel = {
    minValue: 100,
    maxValue: 400,
    options: {
      floor: 0,
      ceil: 500,
      translate: (value: number): string => {
        return '$' + value;
      },
      combineLabels: (minValue: string, maxValue: string): string => {
        return 'from ' + minValue + ' up to ' + maxValue;
      }
    }
  };

  customLabelTemplatesSlider: RangeSliderModel = {
    minValue: 100,
    maxValue: 400,
    options: {
      floor: 0,
      ceil: 500
    }
  };

  alphabetSlider: SimpleSliderModel = {
    value: this.letterToIndex('E'),
    options: {
      stepsArray: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('').map((letter: string): CustomStepDefinition => {
        return { value: this.letterToIndex(letter) };
      }),
      translate: (value: number, label: LabelType): string => {
        return this.indexToLetter(value);
      }
    }
  };

  dateSlider: SimpleSliderModel = {
    value: this.createDateRange()[0].getTime(),
    options: {
      stepsArray: this.createDateRange().map((date: Date) => {
        return { value: date.getTime() };
      }),
      translate: (value: number, label: LabelType): string => {
        return new Date(value).toDateString();
      }
    }
  };

  ticksSlider: SimpleSliderModel = {
    value: 5,
    options: {
      floor: 0,
      ceil: 10,
      showTicks: true
    }
  };

  intermediateTicksSlider: SimpleSliderModel = {
    value: 55,
    options: {
      floor: 0,
      ceil: 100,
      showTicks: true,
      tickStep: 10
    }
  };

  ticksArraySlider: SimpleSliderModel = {
    value: 55,
    options: {
      floor: 0,
      ceil: 100,
      ticksArray: [0, 10, 25, 50, 100]
    }
  };

  ticksLegendSlider: SimpleSliderModel = {
    value: 5,
    options: {
      showTicksValues: true,
      stepsArray: [
        {value: 1, legend: 'Very poor'},
        {value: 2},
        {value: 3, legend: 'Fair'},
        {value: 4},
        {value: 5, legend: 'Average'},
        {value: 6},
        {value: 7, legend: 'Good'},
        {value: 8},
        {value: 9, legend: 'Excellent'}
      ]
    }
  };

  ticksCustomLegendFunctionSlider: SimpleSliderModel = {
    value: 5,
    options: {
      floor: 0,
      ceil: 10,
      showTicks: true,
      getLegend: (value: number): string => {
        return 'T' + value;
      }
    }
  };

  ticksAndTooltipsSlider: SimpleSliderModel = {
    value: 5,
    options: {
      floor: 0,
      ceil: 10,
      showTicks: true,
      ticksTooltip: (v: number): string => {
        return 'Tooltip for ' + v;
      }
    }
  };

  ticksValuesAndTooltipsSlider: SimpleSliderModel = {
    value: 5,
    options: {
      floor: 0,
      ceil: 10,
      showTicksValues: true,
      ticksValuesTooltip: (v: number): string => {
        return 'Tooltip for ' + v;
      }
    }
  };

  rangeTicksValuesSlider: RangeSliderModel = {
    minValue: 1,
    maxValue: 8,
    options: {
      floor: 0,
      ceil: 10,
      showTicksValues: true
    }
  };

  rangeIntermediateTicksSlider: RangeSliderModel = {
    minValue: 15,
    maxValue: 85,
    options: {
      floor: 0,
      ceil: 100,
      showTicksValues: true,
      tickStep: 10,
      tickValueStep: 10
    }
  };

  dynamicTickColorSlider: SimpleSliderModel = {
    value: 0,
    options: {
      ceil: 12,
      floor: 0,
      showSelectionBar: true,
      showTicks: true,
      getTickColor: (value: number): string => {
        if (value < 3) {
          return 'red';
        }
        if (value < 6) {
          return 'orange';
        }
        if (value < 9) {
          return 'yellow';
        }
        return '#2AE02A';
      }
    }
  };

  logScaleSlider: SimpleSliderModel = {
    value: 1,
    options: {
      floor: 1,
      ceil: 100,
      logScale: true,
      showTicks: true
    }
  };

  customScaleSlider: SimpleSliderModel = {
    value: 50,
    options: {
      floor: 0,
      ceil: 100,
      step: 10,
      showTicksValues: true,
      customValueToPosition: (val: number, minVal: number, maxVal: number): number => {
        val = Math.sqrt(val);
        minVal = Math.sqrt(minVal);
        maxVal = Math.sqrt(maxVal);
        const range: number = maxVal - minVal;
        return (val - minVal) / range;
      },
      customPositionToValue: (percent: number, minVal: number, maxVal: number): number => {
        minVal = Math.sqrt(minVal);
        maxVal = Math.sqrt(maxVal);
        const value: number = percent * (maxVal - minVal) + minVal;
        return Math.pow(value, 2);
      }
    }
  };

  draggableRangeSlider: RangeSliderModel = {
    minValue: 1,
    maxValue: 8,
    options: {
      floor: 0,
      ceil: 10,
      draggableRange: true
    }
  };

  draggableRangeOnlySlider: RangeSliderModel = {
    minValue: 4,
    maxValue: 6,
    options: {
      floor: 0,
      ceil: 10,
      draggableRangeOnly: true
    }
  };

  disabledSliderCheckbox: boolean = true;

  disabledSlider: RangeSliderModel = {
    minValue: 20,
    maxValue: 80,
    options: {
      floor: 0,
      ceil: 100,
      step: 10,
      disabled: true,
      showTicks: true,
      draggableRange: true
    }
  };

  readOnlySliderCheckbox: boolean = true;

  readOnlySlider: SimpleSliderModel = {
    value: 50,
    options: {
      floor: 0,
      ceil: 100,
      readOnly: true
    }
  };

  verticalSlider1: SimpleSliderModel = {
    value: 0,
    options: {
      floor: 0,
      ceil: 10,
      vertical: true
    }
  };

  verticalSlider2: RangeSliderModel = {
    minValue: 20,
    maxValue: 80,
    options: {
      floor: 0,
      ceil: 100,
      vertical: true
    }
  };

  verticalSlider3: SimpleSliderModel = {
    value: 5,
    options: {
      floor: 0,
      ceil: 10,
      vertical: true,
      showTicks: true
    }
  };

  verticalSlider4: RangeSliderModel = {
    minValue: 1,
    maxValue: 5,
    options: {
      floor: 0,
      ceil: 6,
      vertical: true,
      showTicksValues: true
    }
  };

  verticalSlider5: SimpleSliderModel = {
    value: 50,
    options: {
      floor: 0,
      ceil: 100,
      vertical: true,
      showSelectionBar: true
    }
  };

  verticalSlider6: SimpleSliderModel = {
    value: 6,
    options: {
      floor: 0,
      ceil: 6,
      vertical: true,
      showSelectionBar: true,
      showTicksValues: true,
      ticksValuesTooltip: (v: number): string => {
        return 'Tooltip for ' + v;
      }
    }
  };

  createDateRange(): Date[] {
    const dates: Date[] = [];
    for (let i: number = 1; i <= 31; i++) {
      dates.push(new Date(2016, 7, i));
    }
    return dates;
  }

  indexToLetter(index: number): string {
    return 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'[index];
  }

  letterToIndex(letter: string): number {
    return 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.indexOf(letter);
  }

  /* TODO: For better or worse, Angular 2+ does not run deep checks on change detection, unlike $watch in angularjs
     The only sane workaround for now is creating a new options object for every update that needs to be done.
     Otherwise, it's a long way down the rabbithole of costly custom check detection
     In future refactoring, I think it's better to move away from nested options to simply a list of @Inputs */
  onChangeDisabledSliderCheckbox(): void {
    this.disabledSlider.options = Object.assign({}, this.disabledSlider.options, {disabled: this.disabledSliderCheckbox});
  }

  onChangeReadOnlySliderCheckbox(): void {
    this.readOnlySlider.options = Object.assign({}, this.readOnlySlider.options, {readOnly: this.readOnlySliderCheckbox});
  }
}
